### PR TITLE
Introduce a pipeline to use the HPC architecture

### PR DESCRIPTION
**Pull request by poquirion**
Using distribute2mp for sct_pipeline so parallel computing is the same on localhost and HPC

### CODE REVIEW

**Comment by poquirion** [comment]
I have also added "Status Legend - " in the report output, it is less confusing that "Status :" alone.

File: scripts/sct_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+from time import time, strftime
+
+import distribute2mpi
+import pandas as pd
+
+import sct_utils as sct
+from msct_parser import Parser
+
+# get path of the toolbox
+# TODO: put it back below when working again (julien 2016-04-04)
```
Comment (peristeri) [inline scripts/sct_pipeline.py]: Is this TODO still valid?

Comment (jcohenadad) [inline scripts/sct_pipeline.py]: i guess not

**Comment by poquirion** [comment]
Tested on guillimin (centos HPC) /bireli (debian) and my laptop (ubuntu).

Need to install the sct with the new --mpi (or -p) option.

File: scripts/msct_shape.py [hunk 2]
```diff
@@ -387,7 +387,7 @@ def prepare_data():
     import isct_test_function
     import json
     json_requirements = 'gm_model=0'
-    data_subjects, subjects_name = isct_test_function.generate_data_list(folder_dataset, json_requirements=json_requirements)
+    data_subjects, subjects_name = sct_pipeline.generate_data_list(folder_dataset, json_requirements=json_requirements)
```
Comment (peristeri) [inline scripts/msct_shape.py]: Where's the import to this module?

**Comment by jcohenadad** [comment]
I've tested version ```poq_issue_1181/199eeb7cf4d4147100609658e46181832a11bb04```, and I see a problem with the passing of output in case of error happening in the test_XXX file. Example below:

~~~
sct_pipeline -f sct_propseg -d ~/sct_example_data/ -p \"-i t2.nii.gz -c t2\"

--
Spinal Cord Toolbox (poq_issue_1181/199eeb7cf4d4147100609658e46181832a11bb04)
Running /Users/julien/code/sct/scripts/sct_pipeline.py -f sct_propseg -d /Users/julien/sct_example_data/ -p "-i t2.nii.gz -c t2"
Check folder existence...
Testing started on: 2017-06-05 13:05:24
SCT commit/branch: 199eeb7cf4d4147100609658e46181832a11bb04/poq_issue_1181
OS: osx (Darwin-14.5.0-x86_64-i386-64bit)
Hostname: julien-macbook.neuro.polymtl.ca
CPU cores: 8
RAM: Primary memory available: 16.00 gigabytes

Command: "sct_propseg -i t2.nii.gz -c t2
Dataset: /Users/julien/sct_example_data/

GLOBAL RESULTS:
Duration: 0s
Passed: 0/4
Crashed: 4/4
Mean: {'dice_segmentation': nan}
STD: {'dice_segmentation': nan}

DETAILED RESULTS:
  subject  dice_segmentation  status
0    dmri                NaN       1
1      mt                NaN       1
2      t1                NaN       1
3      t2                NaN       1
Status Legend - 0: Passed | 1: Crashed | 99: Failed | 200: Input file(s) missing | 201: Ground-truth file(s) missing
~~~

In this example, the following error should be displayed:
~~~
ERROR: when extracting the contrast folder from input file in command line: t2.nii.gz for XXX
~~~

**Comment by jcohenadad** [comment]
@poquirion: when testing on Compute Canada, how many CPU should you ask for in the qsub? Can you please send us a dummy example for the qsub script?

**Comment by poquirion** [comment]
@jcohenadad 
You can ask for N+1 cpu for the N cores requested by sct_pipeline. That is, on guillimin
```
[...]
#PBS -l procs=N+1
[...]
sct_pipeline  -f sct_propseg -d data  -p  \"-i t2s/t2s.nii.gz  -c t2\" -cpu-nb N
```

**Comment by jcohenadad** [comment]
@poquirion:

I've tried this:
~~~
#!/bin/bash
#PBS -l procs=N+1
#PBS -A rrp-355-aa
#PBS -m bea
#PBS -M julien.cohenadad@gmail.com
#PBS -o output.txt
#PBS -e error.txt
export PATH_DATA="/home/jcohen/data/sct_test_function/"
cd $SCRATCH
sct_pipeline -f sct_propseg -d $PATH_DATA -p \"-i t2/t2.nii.gz -c t2\" -cpu-nb N
~~~

It gave me this:
~~~
qsub: submit error (Illegal attribute or resource value for Resource_List.procs)
~~~

Any idea what went wrong?
Thx

**Comment by poquirion** [comment]
Well N should be a number ;)! Let say half or a fourth the numbers of subjects you have in $PATH_DATA.

**Comment by jcohenadad** [comment]
oooops... that's why i asked for a dummy qsub batch --> for dummy people like me ;-)

anyways, it worked, but i am wondering what the "CPU cores: 12" refers to with the mpi activated, since it doesn't seem to match the asked number of cores (8 here). Here is my qsub batch:
~~~
#!/bin/bash
#PBS -l procs=9
#PBS -A rrp-355-aa
#PBS -m bea
#PBS -o output.txt
#PBS -e error.txt
export PATH_DATA="/home/jcohen/data/sct_test_function/"
cd $SCRATCH
sct_pipeline -f sct_propseg -d $PATH_DATA -p \"-i t2/t2.nii.gz -c t2\" -cpu-nb 8
~~~

And here is the result:
~~~
Spinal Cord Toolbox (poq_issue_1181/182b5d9e3c01c609e7325ff2f0ac551374b5fdde)
Sourcing sct_laucher in your environement /home/jcohen/sct
Check folder existence...
Testing started on: 2017-06-06 16:27:40
SCT commit/branch: 182b5d9e3c01c609e7325ff2f0ac551374b5fdde/poq_issue_1181
OS: linux (Linux-2.6.32-642.13.1.el6.x86_64-x86_64-with-centos-6.6-Final)
Hostname: sw-2r10-n31
CPU cores: 12
RAM: MemTotal:       37003668 kB
total       used       free     shared    buffers     cached
Mem:         36136       9591      26545          0         88       1784
-/+ buffers/cache:       7718      28418
Swap:        42711         76      42635

Command: "sct_propseg -i t2/t2.nii.gz -c t2
Dataset: /home/jcohen/data/sct_test_function/

GLOBAL RESULTS:
Duration: 72s
Passed: 3/5
Crashed: 1/5
Mean: {'dice_segmentation': 0.8974604931058585, 'duration [s]': 54.27739925384522}
STD: {'dice_segmentation': 0.08983015356685584, 'duration [s]': 7.992110587842794}

DETAILED RESULTS:
             subject  dice_segmentation  duration [s]  status
0         200_005_s2           0.945733     59.337154       0
1    20150910_joshua           0.762903     52.957869      99
2           amu_G1_6                NaN           NaN     200
3           amu_G1_7                NaN           NaN     200
4           amu_G1_8                NaN           NaN     200
5           bwh_CIS1                NaN           NaN     200
6           bwh_CIS2                NaN           NaN     200
7           errsm_42           0.945600     59.769126       0
8            sct_002           0.935606     58.475603       0
9            sct_004                NaN           NaN     201
10           sct_005                NaN           NaN     201
11           sct_006                NaN           NaN     201
12     vanderbilt_05                NaN     40.847244       1
13  xuanwu_chenxi_01                NaN           NaN     200
Status Legend - 0: Passed | 1: Crashed | 99: Failed | 200: Input file(s) missing | 201: Ground-truth file(s) missing
~~~

**Comment by poquirion** [comment]
The `CPU cores: 12` is just the number of core on the machine where the master process is started. On Guillimin this number will always be 12 or 16, no matter how many core you request. We should perhaps not print that entry in mpi mode.

**Comment by jcohenadad** [comment]
indeed, we should not print that in mpi mode (or print the correct number of cores used)

**Comment by jcohenadad** [comment]
i will approve this PR, but please address those two points in the future:
- systematic redirection of "output" in case of error, and creation of subject-specific log file (done at the parent process instead of within testing) 
- display the correct CPU core in case using --mpi
- doc
- add template files for easy testing

**Pull request merged**